### ISSUE TITLE
Extract list of file/line numbers from script

### ISSUE THREAD

**Issue by rocky** [att/ast#950]
**Description of problem:**

This isn't so much a problem as a feature. In writing a debugger or a profiler it would be helpful to have a list of all the tracable line numbers in a shell script. 

For a debugger, this would allow a check when a breakpoint is requested.
For a profiler, this could be used to determine code coverage.

**Comment by krader1961** [comment]
Are you aware of the special variables such as `LINENO`, `.sh.lineno` and `.sh.file`? It's possible they are broken but, at least on paper, they should address your need.

**Comment by rocky** [comment]
I am aware of both. Those give values that are currently getting run. I wasnt the set of all possible sh.lineno's for a given .sh.file.

**Comment by dannyweldon** [comment]
The `ksh` `-R` option that generates a cross reference database might be good enough if you are just after valid line numbers.  While I don't fully understand the format, a quick look seems to be that
lines starting with `-` are hashes that correspond to `v` variables, `f` files and `p` programs?
Then lines starting with `p` give you the filename hash, type, item hash and line numbers and other info that I don't yet know the meaning of.

**Comment by rocky** [comment]
The following seems to work, even if inelegant:

```shell
function breakpoint_lines {
    typeset -r ksh_file=$1
    typeset -r tempfile=/tmp/ksh-R-$$
    ksh -R $tempfile $ksh_file 2>/dev/null
    grep '^p' $tempfile | cut -d ';' -f 5 | sort -n | uniq
}

breakpoint_lines /etc/profile
```

Here are a couple of things that seem inelegant

* `-R` doesn't allow `-` so it can be put in a pipe
* The invocation of `ksh` is vague. Is there a way to get the ksh executable path? 
* This assumes POSIX utilities grep, sort, unique

I'll leave open in case there is a better way. But feel free to close any time.

**Comment by rocky** [comment]
Actually, when I try this on this code:

```
#!/usr/bin/env ksh
# Note: no CVS Id line since it would mess up regression testing.
# This code is used for various debugger testing.

fn1() {
    echo "fn1 here"
    x=5
    fn3
}

fn2() {
    name="fn2"
    echo "$name here"
    x=6
}

fn3() {
    name="fn3"
    x=$1
}

x=22
y=23
for i in 0 1 3 ; do
  ((x += i))
done
x=27
y=b
x=29
echo $(fn3 30)
fn3 31
fn1;
fn3 33
# source example/dbg-test1.sub
exit 0;
```
Line 25 ``((x += i))`` is missing

**Comment by rocky** [comment]
Yes, I want every line that `set -x` tracing could report excluding those lines that it can't possibly report. LInes with comments, or that are completely blank, or lines that form the interior of strings would never be reported in a `set -x` trace. 

I gave you two use cases and that should have been sufficient, but here is an other example:

```
# This is line 1
x=1
i=2
# This is line 4
((x += i))
# This is line 6
```

Set -x tracing reports: 
```
-(:2):  -[3,, 0]
x=1
-(:3):  -[3,, 0]
i=2
-(:5):  -[3,, 0]
((x += i))
```
so the set of numbers in this example is {2, 3, 5}. The line number 5 is nowhere mentioned via `ksh -R`.

Addiing a flag that outputs 2 3 5 when given a the path a a ksh file would be sufficient.

**Comment by dannyweldon** [comment]
> Given that this file format is completely undocumented my preference is to remove that mention in the ksh man page. At least until such time as we decide to document the output format. Otherwise remove the associated code.

@krader1961, I'm willing to reverse engineer the output format and provide a PR for the man page if you would consider accepting it.

**Comment by rocky** [comment]
@dannyweldon I'd appreciate it if you'd also look into why `(( expression ))` is not showing up. Thanks.a
